### PR TITLE
chore: add docker-compose and dev environment

### PR DESCRIPTION
**Pull request by Divlo**
### What type of change does this PR introduce?
- [ ] Bugfix
- [ ] Feature
- [ ] Refactor
- [ ] Documentation
- [x] Not Sure?

### Does this PR introduce breaking changes?
- [ ] Yes
- [x] No

### List any relevant issue numbers:

Close #164 

### Description:

- [x] Add `docker-compose.yml`
- [x] Add `docker-compose.dev.yml`
- [x] Improve `Dockerfile` : better caching of `npm install` and `pipenv install`

### CODE REVIEW

**Comment by Divlo** [comment]
[🔧 [1.0.0-beta.4] Improve Docker configuration files](https://trello.com/c/Z4flXqAZ/354-%F0%9F%94%A7-100-beta4-improve-docker-configuration-files)

File: README.md [hunk 1]
```diff
@@ -113,12 +113,9 @@ npm start
 # Build
 npm run docker:build
 
-# Run on Linux or macOS
+# Run
 npm run docker:run
 
-# Run on Windows (you can replace "UTC" by your time zone)
-docker run -e TZ=UTC -p 1337:1337 -it leonai/leon
-
```
Comment (louistiti) [inline README.md:121]: Did you try to run it with a Windows OS to see if it breaks without the `TZ` environment variable?

File: Dockerfile [hunk 2]
```diff
@@ -2,25 +2,21 @@ FROM node:14-alpine
 ENV IS_DOCKER true
 WORKDIR /app
 
-# Install system packages
```
Comment (louistiti) [inline Dockerfile:5]: Please keep the comment.

File: Dockerfile [hunk 3]
```diff
@@ -2,25 +2,21 @@ FROM node:14-alpine
...
 RUN apk add --no-cache --no-progress \
     ca-certificates \
     python3 \
     git \
     tzdata
 
-# Upgrade pip and install Pipenv
```
Comment (louistiti) [inline Dockerfile:12]: Please keep the comment.

File: package.json [hunk 4]
```diff
@@ -47,8 +47,9 @@
     "prepare-release": "babel-node scripts/release/prepare-release.js",
     "check": "babel-node scripts/run-check.js",
     "docker:build": "docker build -t leonai/leon .",
-    "docker:run": "docker run -e TZ=`cat /etc/timezone` -p 1337:1337 -it leonai/leon",
```
Comment (louistiti) [inline package.json:50]: Did you try to run it with a Windows OS to see if it breaks without the `TZ` environment variable?

File: README.md (hunk 1 above)
Comment (Divlo) [reply README.md:121]: To be honest, no.
Thanks for this catch, I'll test if it's work, if not, I would like that it stays straight forward, doesn't matter the OS.

Yeah, I'll look at it, and will try to understand the issue. :+1:

File: Dockerfile (hunk 2 above)
Comment (Divlo) [reply Dockerfile:5]: We should avoid useless comments, it's quite obvious that we're installing system packages, with `apt add`.

Do you see any case, where this comment could be useful ?
If you really want, I could add it again, sure, just wanted to know if there is a reason to keep it. :smile:

File: Dockerfile (hunk 3 above)
Comment (Divlo) [reply Dockerfile:12]: Same as `apk add`, it is obvious.

Do you see any case where this is useful ?

Sure, I could add it again, if you really want to.

File: Dockerfile [hunk 5]
```diff
@@ -2,25 +2,21 @@ FROM node:14-alpine
...
 RUN pip3 install --no-cache-dir --progress-bar off pipenv
 
-COPY . .
-
-# Install Leon
-# Need to explicitly run the npm preinstall and npm posinstall scripts
-# because npm tries to downgrade its privileges, and these scripts are not executed
```
[the remark this answers is not in the document]
Comment (Divlo) [reply Dockerfile:19]: Right! I'll add it again. :+1:

Please note that wiith `npm@7`, we don't anymore to run these npm scripts like `preinstall` in a Docker container, the issue has been fixed, in the latest major version of `npm`.

File: README.md (hunk 1 above)
Comment (louistiti) [reply README.md:121]: Awesome! Looking forward to see your feedback :+1:

File: Dockerfile (hunk 2 above)
Comment (louistiti) [reply Dockerfile:5]: Generally speaking I like to keep things divided into their own chunk with a short description of what is going next. Also, please keep in mind that not everyone is familiar with UNIX commands or whatsoever. So yes, please add them back.

File: Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (louistiti) [reply Dockerfile:19]: Oh good to know, thanks! Could you please share a link for the reference?

[the remark this answers is not in the document]
Comment (Divlo) [reply Dockerfile:19]: Actually, it is just something I noticed, I don't really read the changelogs about `npm`.
Not to say nonsense, I'll test it again with npm@7 and see if it is indeed the case, but I'm pretty sure it is, maybe I'm mistaken, will stay you inform.

[the remark this answers is not in the document]
Comment (Divlo) [reply Dockerfile:19]: With this Dockerfile :
```Dockerfile
FROM node:14-alpine
RUN npm install --global npm@7
ENV IS_DOCKER true
WORKDIR /app

RUN apk add --no-cache --no-progress \
    ca-certificates \
    python3 \
    git \
    tzdata

RUN pip3 install --no-cache-dir --progress-bar off pipenv

COPY ./ ./
RUN npm clean-install
RUN pipenv install
RUN npm run build

CMD ["npm", "start"]
```

You can see that it is running the `preinstall` script, even if I don't specify it.
image

[the remark this answers is not in the document]
Comment (louistiti) [reply Dockerfile:19]: Awesome! Looks good for this chunk then :slightly_smiling_face:

**Review by louistiti — APPROVED** [review]
LGTM. Awesome work, thanks! :tada:

**Pull request merged**